### PR TITLE
Rspec  functional testing for User model

### PR DESCRIPTION
**Pull request by sonishc**
#### Unit-tests 
Project
  - Valid?
    - valid with valid attributes
    - invalid without name attribute

User
  - Valid?
    - valid with valid attributes
  - .filter
    - call without param
      - returns all users
    - when method with id Project.first
      - returns users list by Project.first.id
  - .as_json
    - when had projects
      - returns an array of objects as json
  - .from_omniauth
    - when user use Google SSO
      - returns not to change User count
      - returns new google user that ready to sign up

### CODE REVIEW

File: spec/models/project_spec.rb [hunk 1]
```diff
@@ -1,5 +1,10 @@
 require 'rails_helper'
 
 RSpec.describe Project, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  let(:project) { build(:project) }
+  context 'associations' do
+    it 'has_many users' do
+      puts project.users.inspect
```
Comment (rubycop) [inline spec/models/project_spec.rb]: What this test does?
We use `puts` only for debuging

File: spec/factories/users.rb [hunk 2]
```diff
@@ -2,7 +2,10 @@
   factory :user do
     last_name { 'MyString' }
     first_name { 'MyString' }
-    email { 'MyString' }
+    email { 'MyString@MyString' }
```
Comment (rubycop) [inline spec/factories/users.rb]: It should be more realistic email like `user@example.com`

File: spec/factories/users.rb [hunk 3]
```diff
@@ -2,7 +2,10 @@
...
     phone { 'MyString' }
```
Comment (rubycop) [inline spec/factories/users.rb]: Also phone field

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -1,5 +1,74 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  let(:user) { build(:user) }
+  let!(:next_user) { create(:user) }
+  let(:attributes_presence) { %i[first_name last_name phone project_id email] }
+
+  context 'when got project id' do
+    it 'should get error message if id absent' do
```
Comment (rubycop) [inline spec/models/user_spec.rb]: instead of `should get...` use `gets...`

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -1,5 +1,74 @@
...
+      id = 22_423
+      expect(Project.exists?(id)).to eql(false)
```
Comment (rubycop) [inline spec/models/user_spec.rb]: Where is your expected error message?
Your title doesn't match your expectation.

Comment (rubycop) [reply spec/models/user_spec.rb]: You shouldn't use random id to check user's id. It totally wrong test

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -1,5 +1,74 @@
...
+    end
+
+    it 'should have associated projects via id' do
+      id = 2
```
Comment (rubycop) [inline spec/models/user_spec.rb]: Also here.. You shouldn't have use id.
Use your factories to operate with data in test.

File: spec/models/project_spec.rb [hunk 7]
```diff
@@ -1,5 +1,10 @@
 require 'rails_helper'
 
 RSpec.describe Project, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  let(:project) { build(:project) }
```
Comment (rubycop) [inline spec/models/project_spec.rb]: use factory (valid_project, invalid_project)

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -1,5 +1,74 @@
...
+      projects = Project.find(id).users
+      users = User.filter(id)
```
Comment (rubycop) [inline spec/models/user_spec.rb]: Why you use filter here?

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -1,5 +1,74 @@
...
+      expect(users.count).to eql(projects.count)
+    end
+    it 'should have all projects via id 0' do
```
Comment (rubycop) [inline spec/models/user_spec.rb]: Wrong test

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -1,5 +1,74 @@
...
+      id = 0
+      projects = Project.all
+      users = User.filter(id)
+      expect(users.count).to eql(projects.count)
+    end
+    it 'sdsadasdsaa' do
```
Comment (rubycop) [inline spec/models/user_spec.rb]: The same

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -1,5 +1,74 @@
...
+      users = User.all
+      User.as_json(users)
+    end
+  end
+
+  context 'validations uniqueness' do
```
Comment (rubycop) [inline spec/models/user_spec.rb]: We don't need this context. Please remove

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -1,5 +1,74 @@
...
+      user.valid?
+      expect(user.errors[:email]).not_to include(error_uniqueness_message)
+    end
+    it "hasn't a unique phone" do
+      user.phone = '380(63)-333-33-33'
+      user.valid?
+      expect(user.errors[:phone]).not_to include(error_uniqueness_message)
+    end
+  end
+  context 'validations presence :attributes_presence' do
```
Comment (rubycop) [inline spec/models/user_spec.rb]: And this

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -1,5 +1,74 @@
...
+    end
+    it 'is not valid presence :attributes_presence' do
+      attributes_presence.each do |attr|
+        user[attr] = ''
+        user.valid?
+        expect(user.errors[attr]).to include(error_presence_message)
+      end
+    end
+  end
+  context 'associations' do
```
Comment (rubycop) [inline spec/models/user_spec.rb]: Wrong test w/o expectation block

**Comment by rubycop** [comment]
> #### Validations-test
> * presence
> * uniqueness
> 
> _I want to know. Is it correct sequence ?_

We don't need test rails stuff. We trust rails and know that rails does it well.
Instead of this we write test for our data and check them.
For example we should write a test that checks what fields should take a part in validation.

File: spec/factories/users.rb [hunk 14]
```diff
@@ -1,8 +1,11 @@
...
-    first_name { 'MyString' }
-    email { 'MyString' }
-    phone { 'MyString' }
+    last_name { 'Last_Name' }
+    first_name { 'First_Name' }
+    email { 'user@example.org' }
+    phone { '+380000000000' }
+    password { 'secret' }
+    password_confirmation { password }
+    project_id { 1 }
```
Comment (afurm) [inline spec/factories/users.rb]: Do not hard code the IDs. Read about association creating here: https://qiita.com/Kolosek/items/7c85337f928161e0e76e

Comment (sonishc) [reply spec/factories/users.rb]: > Do not hard code the IDs. Read about association creating here: https://qiita.com/Kolosek/items/7c85337f928161e0e76e

And this factory was merged with a master branch a few hours ago.
Good, I'll change it.

File: spec/rails_helper.rb [hunk 15]
```diff
@@ -1,6 +1,22 @@
 # This file is copied to spec/ when you run 'rails generate rspec:install'
 require 'simplecov'
+require 'omniauth'
 SimpleCov.start
+
+OmniAuth.config.test_mode = true
```
Comment (VasylYaremkiv) [inline spec/rails_helper.rb]: Move to `spec/support/google_auth_stub.rb`

File: spec/models/user_spec.rb [hunk 16]
```diff
@@ -1 +1,83 @@
...
+RSpec.describe User, type: :model do
+  describe 'Valid?' do
+    let(:user) { build(:user) }
+    it 'valid with valid attributes' do
+      expect(user).to be_valid
+    end
+  end
+
+  describe '.filter' do
+    subject { described_class.filter(id) }
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: Do not use `subject` with parameter.

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -1 +1,83 @@
...
+    let!(:users) { 5.times { create(:user) } }
+
+    context 'when argument eq 0' do
+      let(:id) { 0 }
```
[the remark this answers is not in the document]
Comment (sonishc) [reply spec/models/user_spec.rb]: Why not used?  It It used in subject { described_class.filter(**id**) }

File: spec/models/user_spec.rb [hunk 18]
```diff
@@ -1 +1,82 @@
...
+    context 'call without param' do
+      it 'returns all users' do
+        expect(User.filter(nil)).to eq User.includes(:project)
+      end
+    end
+
+    context 'when method with id Project.first' do
+      it 'returns users list by Project.first.id' do
+        id = Project.first.id
+        expect(User.filter(id)).to eq equal_for_filter(id)
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]:
```
expected_user_ids = [foo_user.id, bar_user_id]
expect(User.filter(id).map(&:id)).to eq(expected_user_ids)
```

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -1 +1,82 @@
...
+
+  describe '.from_omniauth' do
+    subject { described_class.from_omniauth(access_token) }
+    context 'when user use Google SSO' do
+      let(:access_token) { OmniAuth.config.mock_auth[:google] }
+      let(:user_valid) { create(:user, email: 'test@google.com') }
+
+      it 'returns created user' do
+        user_valid
+        user = User.find_by(email: access_token[:info][:email])
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: check if User.count does not change

File: spec/models/user_spec.rb [hunk 20]
```diff
@@ -1 +1,82 @@
...
+      end
+
+      it 'returns not valid new user' do
+        data = access_token[:info]
+        user = User.new(
+          first_name: data[:name].partition(' ').first,
+          last_name: data[:name].partition(' ').last,
+          email: data[:email], project: Project.first
+        )
+        user[:encrypted_password] = subject[:encrypted_password]
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: check if User.count change
Retrieve user by email and check fields (first_name and last_name)

File: spec/rails_helper.rb [hunk 21]
```diff
@@ -19,4 +22,5 @@
   config.infer_spec_type_from_file_location!
   config.filter_rails_from_backtrace!
   config.include Devise::Test::ControllerHelpers, type: :controller
+  config.include FactoryBot::Syntax::Methods
```
Comment (VasylYaremkiv) [inline spec/rails_helper.rb]: Dir[Rails.root.join("spec/support/**/*.rb")].sort.each {|f| require f}

File: spec/factories/projects.rb [hunk 22]
```diff
@@ -1,6 +1,13 @@
 FactoryBot.define do
   factory :project do
-    name { 'ProjectName' }
-    desc { 'ProjectDescription' }
+    sequence(:name)      { |n| "ProjectName#{n}" }
+    sequence(:desc)      { |n| "ProjectDescription#{n}" }
+
+    trait :invalid_attributes do
```
Comment (VasylYaremkiv) [inline spec/factories/projects.rb]: remove invalid factory

**Comment by sonishc** [comment]
@VasylYaremkiv  I am ready to new review

File: spec/models/user_spec.rb [hunk 23]
```diff
@@ -1 +1,74 @@
...
+    let(:user_john) { create(:user, first_name: 'user_john', project: second_project) }
+
+    context 'call without param' do
+      it 'returns all users' do
+        expected_all_user_ids = [user_foo.id, user_bar.id, user_john.id]
+        expect(User.filter(nil).map(&:id)).to eq expected_all_user_ids
+      end
+    end
+
+    context 'when method with id Project.first' do
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: Do not use `Project.first` in test name

File: spec/models/user_spec.rb [hunk 24]
```diff
@@ -1 +1,74 @@
...
+    end
+  end
+
+  describe '.from_omniauth' do
+    context 'when user use Google SSO' do
+      let(:access_token) { OmniAuth.config.mock_auth[:google] }
+      let(:registered_google_user) { create(:user, email: access_token[:info][:email]) }
+      let(:new_user) { build(:user, email: access_token[:info][:email]) }
+
+      it 'returns not to change User count' do
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: I do not understand this test. What do you test in?

File: spec/models/user_spec.rb [hunk 25]
```diff
@@ -1 +1,74 @@
...
+        registered_google_user
+        expect { new_user.save }.not_to change(User, :count)
+      end
+
+      it 'returns new google user that ready to sign up' do
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: Use correct name for test

File: spec/models/user_spec.rb [hunk 26]
```diff
@@ -1 +1,74 @@
...
+        name = access_token[:info][:name]
+
+        new_user.first_name = name.partition(' ').first
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: Why do you set some fields for new user?

Comment (sonishc) [reply spec/models/user_spec.rb]: @VasylYaremkiv
These fields simulate Google user fields

File: spec/models/user_spec.rb (hunk 24 above)
Comment (sonishc) [reply spec/models/user_spec.rb]: @VasylYaremkiv 
When google user registered in the application, this test check if returned not changed count of users.

**Comment by sonishc** [comment]
@VasylYaremkiv  I am ready to new review

**Comment by sonishc** [comment]
> #### Unit-tests
>     @sonishc PTAL my comments 

Dear @VasylYaremkiv  
@sonishc waiting for your reviews

**Comment by sonishc** [comment]
> #### Unit-tests
>     Replaced first and last names
>     Added if statments for .from_omniauth
>     Removed unused users names changed to sequence in factory

Comment for @VasylYaremkiv

**Comment by VasylYaremkiv** [comment]
Looks good! Squash commits and PR will be ready for merge.

**Comment by sonishc** [comment]
> Looks good! Squash commits and PR will be ready for merge.
@VasylYaremkiv 
Commits are squashed

**Pull request merged**